### PR TITLE
`k6 login cloud` command

### LINKED ISSUE #367: Cloud collector login command
Implement a `k6 cloud login` command to allow users to login to their Load Impact account so that streaming of results or cloud executed tests will end up under the correct organization and project.

The input prompt presented to user should say:
`Email or token:`

and accept either a pasted API token or an email address (look for @ sign?) in which case it should ask for the password:

`Password:`

The result stored in the config file should always be the API token (not the email + password).

### PR DESCRIPTION
**Pull request by ppcano**
Will close #367

`k6 login cloud` has three options:

-  `k6 login cloud -s`: show the stored token.
- `k6 login cloud -t YOUR_TOKEN`: set up the token.
- `k6 login cloud`:  ask for your Load Impact user email and password to automatically set up the token.

@liclac 

From a functional perspective, it looks ready.

Some possible next steps:
- hidden the password. Found this [post](https://stackoverflow.com/questions/2137357/getpasswd-functionality-in-go). Thoughts?
- replace the `login` request for a new endpoint request to avoid the error: `You have to create an API Token with your Load Impact account.`

### CODE REVIEW

File: cmd/login_cloud.go [hunk 1]
```diff
@@ -1,50 +1,111 @@
...
+			client := cloud.NewClient("", conf.Host, Version)
+			response, err := client.Login(email, password)
+			if err != nil {
+				return errors.New("Failed to login.")
+			}
+
+			if response.APIToken == "" {
+				// TODO: instead of `Login`, we must create an endpoint `GetorCreateAPIToken`:
+				//       Given an email and password, it will return your Token or create a new one.
+				return errors.New("You have to create an API Token with your Load Impact account.")
```
Comment (liclac) [inline cmd/login_cloud.go]: Maybe reword this a little to sound less accusatory?

"Your account has no API token, please generate one: https://loadimpact.com/whichever/page/you/do/that/on", something like that.

File: cmd/login_cloud.go [hunk 2]
```diff
@@ -1,50 +1,111 @@
...
+			if err != nil {
+				return err
+			}
+
+			email := vals["Email"].(string)
+			password := vals["Password"].(string)
+			client := cloud.NewClient("", conf.Host, Version)
+			response, err := client.Login(email, password)
+			if err != nil {
+				return errors.New("Failed to login.")
```
Comment (liclac) [inline cmd/login_cloud.go]: Bad. Return the error. Think of the bug reports.

File: cmd/login_cloud.go [hunk 3]
```diff
@@ -1,50 +1,111 @@
...
+	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
 )
 
+var (
+	token string
+	show  bool
+)
+
+func printToken(conf cloud.Config) {
```
Comment (liclac) [inline cmd/login_cloud.go]: You can make functions that are only used in one scope into closure variables.

```go
printToken := func(conf cloud.Config) {
  // ...
}
```

File: cmd/login_cloud.go [hunk 4]
```diff
@@ -1,50 +1,111 @@
...
-		form := ui.Form{
-			Fields: []ui.Field{
-				ui.StringField{
-					Key:     "token",
-					Label:   "API Token",
-					Default: conf.Token,
-				},
-			},
+
+		if show {
```
Comment (liclac) [inline cmd/login_cloud.go]: Instead of all these if's and if-elses, consider using a switch.

File: cmd/login_cloud.go [hunk 5]
```diff
@@ -1,50 +1,111 @@
...
-		return writeDiskConfig(cdir, config)
+		printToken(conf)
+		return nil
+
 	},
 }
 
 func init() {
 	loginCmd.AddCommand(loginCloudCommand)
+	loginCloudCommand.Flags().StringVarP(&token, "token", "t", token, "setup the Load Impact Token")
```
Comment (liclac) [inline cmd/login_cloud.go]: Not very useful as a tooltip, "specify \`token\` to use" or "use an existing \`token\` might be better. (Using \` replaces the type shown for the flag on the --help screen.)

File: cmd/login_cloud.go [hunk 6]
```diff
@@ -1,50 +1,111 @@
...
+	loginCloudCommand.Flags().BoolVarP(&show, "show", "s", false, "show the saved Load Impact Token")
```
Comment (liclac) [inline cmd/login_cloud.go]: Maybe something like "display saved token and exit".

File: cmd/login_cloud.go [hunk 7]
```diff
@@ -1,50 +1,111 @@
...
 
+var (
+	token string
+	show  bool
+)
+
+func printToken(conf cloud.Config) {
+	label := "Token"
+	displayLabel := " " + color.New(color.Faint, color.FgCyan).Sprint("["+label+"]")
+	fmt.Fprintf(stdout, " "+displayLabel+": "+conf.Token+"\n")
```
Comment (liclac) [inline cmd/login_cloud.go]: For consistency, just do

```go
fmt.Fprintf(stdout, " token: %s\n", ui.ValueColor.Sprint(conf.Token))`.
```

Honestly, you can probably restructure the whole function to just end in this without needing a function at all.

File: stats/cloud/api.go [hunk 8]
```diff
@@ -112,3 +116,30 @@ func (c *Client) TestFinished(referenceID string, thresholds ThresholdResult, ta
...
+	req, err := c.NewRequest("POST", url, data)
+	if err != nil {
+		return nil, err
+	}
+
+	lr := LoginResponse{}
+	err = c.Do(req, &lr)
+	if err != nil {
+		// if wrong email or password, it will throw an Error
+		// `client.Do` should handle this in a better way
```
Comment (liclac) [inline stats/cloud/api.go]: How so...? An error seems appropriate.

Also: capital letters at the start of sentences, periods at the end.

**Comment by ppcano** [comment]
@liclac 

Changes are done. I didn't implement the `switch`, but I think, most of them are completed. Feel free to suggest more changes.

> Consider calling some API function to test the token instead of just saving it. Is there one? Most APIs have one.

Thanks. Good idea, I've added it to the backlog. We will also improve or replace the `login` command to auto-generate the `token` if not available.

**Comment by liclac** [comment]
Please do use a switch, because that's an awkward amount of indentation for what is essentially a three-way switch.

**Comment by ppcano** [comment]
@liclac Done fc306f145f9ffc450b0b0776c95cffdf341dfe76

File: cmd/login_cloud.go [hunk 9]
```diff
@@ -10,6 +10,12 @@ import (
 	"github.com/spf13/cobra"
 )
 
+const (
+	actionSetToken = iota
```
Comment (liclac) [inline cmd/login_cloud.go]: This is not a good prefix. Just use `flags.GetBool()` instead.

File: cmd/login_cloud.go [hunk 10]
```diff
@@ -1,50 +1,116 @@
...
+
 		conf := config.Collectors.Cloud
-		form := ui.Form{
-			Fields: []ui.Field{
-				ui.StringField{
-					Key:     "token",
-					Label:   "API Token",
-					Default: conf.Token,
+
+		switch cmdName {
```
Comment (liclac) [inline cmd/login_cloud.go]: Switches don't need a variable.

```go
switch {
case flags.GetBool("show"):
  // ...
case flags.GetBool("token"):
  // …
default:
  // ...
}
```

**Comment by ppcano** [comment]
@liclac 

As far as I know `flags.GetBool` returns multiple values and cannot be used in a `switch` case.

fe15b4ef8e118e6eb6405f9c48e9ee04605ae82c refactors the `switch` to avoid the `const` and `cmdName` variable.

File: stats/cloud/api.go [hunk 11]
```diff
@@ -112,3 +116,28 @@ func (c *Client) TestFinished(referenceID string, thresholds ThresholdResult, ta
...
+
+	req, err := c.NewRequest("POST", url, data)
+	if err != nil {
+		return nil, err
+	}
+
+	lr := LoginResponse{}
+	err = c.Do(req, &lr)
+	if err != nil {
+		return nil, errors.New("Failed to login.")
```
Comment (robingustafsson) [inline stats/cloud/api.go]: @ppcano Isn't this going to mask the error response from the server? Wouldn't it be better to just forward/return the `err` instead of creating a new one that hides the underlying problem.

Related but perhaps to consider separately, we should handle rate-limit error responses. We don't have any checks for 429 status code in `checkResponse` (https://github.com/loadimpact/k6/blob/master/stats/cloud/client.go#L156).

File: cmd/login_cloud.go [hunk 12]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
 	Short: "Authenticate with Load Impact",
 	Long: `Authenticate with Load Impact.
 
-This will set the default server used when just "-o cloud" is passed.`,
+This will set the default token used when just "k6 run -o cloud" is passed.`,
```
Comment (robingustafsson) [inline cmd/login_cloud.go]: I think we should be more descriptive here, maybe:
"By authenticating you tell k6 to store your account API token in a configuration file on disk for subsequent use when running "k6 run -o cloud"."

File: cmd/login_cloud.go [hunk 13]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
+
+	Example: `
+  # Show the stored token.
+  k6 login cloud -s
+
+  # Set up the token.
```
Comment (robingustafsson) [inline cmd/login_cloud.go]: Maybe "Persist the specified API token to configuration file on disk"?

File: cmd/login_cloud.go [hunk 14]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
 	Short: "Authenticate with Load Impact",
 	Long: `Authenticate with Load Impact.
 
-This will set the default server used when just "-o cloud" is passed.`,
+This will set the default token used when just "k6 run -o cloud" is passed.`,
+
+	Example: `
+  # Show the stored token.
```
Comment (robingustafsson) [inline cmd/login_cloud.go]: Maybe "Show the persisted API token from configuration file on disk"?

File: cmd/login_cloud.go [hunk 15]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
+  k6 login cloud -s
+
+  # Set up the token.
+  k6 login cloud -t YOUR_TOKEN
+
+  # Ask for your Load Impact user email and password to automatically set up the token.
```
Comment (robingustafsson) [inline cmd/login_cloud.go]: Maybe "Authenticate using your Load Impact user email and password to automatically persist an API token in configuration file on disk."?

File: cmd/login_cloud.go (hunk 12 above)
Comment (robingustafsson) [reply cmd/login_cloud.go]: ...and this should be updated to mention cloud execution later.

File: stats/cloud/api.go (hunk 11 above)
Comment (ppcano) [reply stats/cloud/api.go]: @robingustafsson This is a temporary solution until we handle better the error message from the server (including the limitations for the `rate`limits). This will be handled as a separate task.

**Comment by ppcano** [comment]
@liclac what do you think about Robin text changes? I have already changed these texts before. 

Please, let's move forward to get this PR merged, I need this for the `cloud` execution.

File: stats/cloud/api.go (hunk 11 above)
Comment (liclac) [reply stats/cloud/api.go]: No masking error messages.

File: cmd/login_cloud.go (hunk 14 above)
Comment (liclac) [reply cmd/login_cloud.go]: I think "Show the stored token" is fine, honestly.

File: cmd/login_cloud.go (hunk 13 above)
Comment (liclac) [reply cmd/login_cloud.go]: Maybe "Store a token"?

File: cmd/login_cloud.go (hunk 15 above)
Comment (liclac) [reply cmd/login_cloud.go]: "Log in with an email/password"?

File: cmd/login_cloud.go [hunk 16]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
 	Short: "Authenticate with Load Impact",
 	Long: `Authenticate with Load Impact.
 
-This will set the default server used when just "-o cloud" is passed.`,
+This will set the default token used when just "k6 run -o cloud" is passed.`,
+
```
Comment (liclac) [inline cmd/login_cloud.go]: No need for this newline.

File: cmd/login_cloud.go [hunk 17]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
+	Example: `
+  # Show the stored token.
+  k6 login cloud -s
+
+  # Set up the token.
+  k6 login cloud -t YOUR_TOKEN
+
+  # Ask for your Load Impact user email and password to automatically set up the token.
+  k6 login cloud`[1:],
+
```
Comment (liclac) [inline cmd/login_cloud.go]: No need for this newline.

File: cmd/login_cloud.go [hunk 18]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
 	Args: cobra.NoArgs,
 	RunE: func(cmd *cobra.Command, args []string) error {
+
```
Comment (liclac) [inline cmd/login_cloud.go]: No need for this newline.

File: cmd/login_cloud.go [hunk 19]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
-				ui.StringField{
-					Key:     "token",
-					Label:   "API Token",
-					Default: conf.Token,
-				},
-			},
-		}
-		vals, err := form.Run(os.Stdin, stdout)
+		flags := cmd.Flags()
+		show, err := flags.GetBool("show")
```
Comment (liclac) [inline cmd/login_cloud.go]: Maybe use `getNullBool()` here and just check `show.Bool` (no need to check `.Valid` if it doesn't matter if it's an explicit default value or unset).

The only reason `GetX()` will return an error is if you get a flag that doesn't exist; `getNullX` will just panic instead, because that's a programmer error.

File: cmd/login_cloud.go [hunk 20]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
-			Fields: []ui.Field{
-				ui.StringField{
-					Key:     "token",
-					Label:   "API Token",
-					Default: conf.Token,
-				},
-			},
-		}
-		vals, err := form.Run(os.Stdin, stdout)
+		flags := cmd.Flags()
```
Comment (liclac) [inline cmd/login_cloud.go]: Just use `cmd.Flags()`.

File: cmd/login_cloud.go [hunk 21]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
+  k6 login cloud`[1:],
+
 	Args: cobra.NoArgs,
 	RunE: func(cmd *cobra.Command, args []string) error {
+
+		printToken := func(conf cloud.Config) {
+			fmt.Fprintf(stdout, "  token: %s\n", ui.ValueColor.Sprint(conf.Token))
+		}
+
 		config, cdir, err := readDiskConfig()
```
Comment (liclac) [inline cmd/login_cloud.go]: Let's move this down below the flag stuff, with the `conf :=` line.

File: cmd/login_cloud.go [hunk 22]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
+  # Set up the token.
+  k6 login cloud -t YOUR_TOKEN
+
+  # Ask for your Load Impact user email and password to automatically set up the token.
+  k6 login cloud`[1:],
+
 	Args: cobra.NoArgs,
 	RunE: func(cmd *cobra.Command, args []string) error {
+
+		printToken := func(conf cloud.Config) {
```
Comment (liclac) [inline cmd/login_cloud.go]: Instead of having this function and calling it three times, put it at the end of the function and nothing in `case show`.

File: cmd/login_cloud.go [hunk 23]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
+
+			vals, err := form.Run(os.Stdin, stdout)
+			if err != nil {
+				return err
+			}
+
+			email := vals["Email"].(string)
+			password := vals["Password"].(string)
+			client := cloud.NewClient("", conf.Host, Version)
+			response, err := client.Login(email, password)
```
Comment (liclac) [inline cmd/login_cloud.go]: `res` or `resp` are our typical variable names for responses.

File: cmd/login_cloud.go [hunk 24]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
+					},
+				},
+			}
+
+			vals, err := form.Run(os.Stdin, stdout)
+			if err != nil {
+				return err
+			}
+
+			email := vals["Email"].(string)
```
Comment (liclac) [inline cmd/login_cloud.go]: Nitpick: I'd group `email` and `password` with the `form.Run()` above instead of with the response stuff.

File: cmd/login_cloud.go [hunk 25]
```diff
@@ -14,37 +16,96 @@ var loginCloudCommand = &cobra.Command{
...
+			password := vals["Password"].(string)
+			client := cloud.NewClient("", conf.Host, Version)
+			response, err := client.Login(email, password)
+			if err != nil {
+				return err
+			}
+
+			if response.APIToken == "" {
+				return errors.New("Your account has no API token, please generate one: `https://app.loadimpact.com/account/token`.")
```
Comment (liclac) [inline cmd/login_cloud.go]: The backticks break clickable links in some terminals :(

**Comment by ppcano** [comment]
@liclac Thanks for looking at this. The new requested changes have been included.

File: cmd/login_cloud.go [hunk 26]
```diff
@@ -14,37 +16,86 @@ var loginCloudCommand = &cobra.Command{
...
+			fmt.Fprintf(stdout, "  token: %s\n", ui.ValueColor.Sprint(conf.Token))
+		}
+		conf := config.Collectors.Cloud
+
+		switch {
+		case show.Bool:
+		case token != "":
+			conf.Token = token
+		default:
+			printToken(conf)
```
Comment (liclac) [inline cmd/login_cloud.go]: Drop this, and the `printToken()` function, just always do the `fmt.Fprintf` call normally at the end.

File: cmd/login_cloud.go [hunk 27]
```diff
@@ -14,37 +16,86 @@ var loginCloudCommand = &cobra.Command{
...
-					Key:     "token",
-					Label:   "API Token",
-					Default: conf.Token,
-				},
-			},
-		}
-		vals, err := form.Run(os.Stdin, stdout)
+		flags := cmd.Flags()
+		show := getNullBool(flags, "show")
+		token, err := flags.GetString("token")
```
Comment (liclac) [inline cmd/login_cloud.go]: Sigh. Use `getNullX()` in **both** places.

File: cmd/login_cloud.go (hunk 20 above)
Comment (liclac) [reply cmd/login_cloud.go]: Still just use `cmd.Flags()` normally in the calls below. This doesn't need a temporary variable.

File: cmd/login_cloud.go [hunk 28]
```diff
@@ -14,37 +16,86 @@ var loginCloudCommand = &cobra.Command{
...
+			password := vals["Password"].(string)
+
+			client := cloud.NewClient("", conf.Host, Version)
+			res, err := client.Login(email, password)
+			if err != nil {
+				return err
+			}
+
+			if res.APIToken == "" {
+				return errors.New("Your account has no API token, please generate one: \"https://app.loadimpact.com/account/token\".")
```
Comment (liclac) [inline cmd/login_cloud.go]: You can actually skip escaping the " here if you use use ` for the outer quotes.

File: stats/cloud/api.go [hunk 29]
```diff
@@ -55,6 +55,10 @@ type CreateTestRunResponse struct {
 	ReferenceID string `json:"reference_id"`
 }
 
+type LoginResponse struct {
+	APIToken string `json:"token"`
```
Comment (liclac) [inline stats/cloud/api.go]: Oh, why not just call this `Token` like the JSON field suggests?

**Comment by ppcano** [comment]
@liclac Changes added for the `k6 login cloud`. I will begin now with the `k6 cloud`.

**Comment by liclac** [comment]
@robingustafsson Just waiting for your review to merge 👍
